### PR TITLE
DTSPO-7161 Enable sync branches with master by default

### PR DESCRIPTION
**Pull request by Tyler-35**
Changes:
Environment branches demo, ithc and perftest have been set to sync to master by default.
readme updated.

### CODE REVIEW

**Review by timja — CHANGES_REQUESTED** [review]

**Comment by timja** [comment]
Link to where you have tested this?

**Comment by Tyler-35** [comment]
> 
Been testing it on this PR https://sds-build.platform.hmcts.net/job/HMCTS/job/sds-toffee-recipes-service/view/change-requests/job/PR-110/

**Comment by timja** [comment]
> > 
> 
> Been testing it on this PR [sds-build.platform.hmcts.net/job/HMCTS/job/sds-toffee-recipes-service/view/change-requests/job/PR-110](https://sds-build.platform.hmcts.net/job/HMCTS/job/sds-toffee-recipes-service/view/change-requests/job/PR-110/)

you would need to test it on master wouldn't you to make sure the syncing is working as you expect/

**Comment by Tyler-35** [comment]
> > > 
> > 
> > 
> > Been testing it on this PR [sds-build.platform.hmcts.net/job/HMCTS/job/sds-toffee-recipes-service/view/change-requests/job/PR-110](https://sds-build.platform.hmcts.net/job/HMCTS/job/sds-toffee-recipes-service/view/change-requests/job/PR-110/)
> 
> you would need to test it on master wouldn't you to make sure the syncing is working as you expect?

I thought by pulling the lastest from master and changing to the sync to run on pr looks to be testing the sync. https://sds-build.platform.hmcts.net/blue/organizations/jenkins/HMCTS%2Fsds-toffee-recipes-service/branches/

File: vars/sectionSyncBranchesWithMaster.groovy [hunk 1]
```diff
@@ -10,32 +10,39 @@
...
-    def branchesToSync = params.branchestoSync
+    def branchesToSync = params.branchestoSync ?: ['ithc', 'demo', 'perftest']
     def product = params.product
     def credentialsId = env.GIT_CREDENTIALS_ID
 
-    if (!branchesToSync.isEmpty()) {
-        stageWithAgent("Sync Branches with Master", product) {
+
+    stageWithAgent("Sync Branches with Master", product) {
+        if (!config.branchesToSyncWithMaster.isEmpty()) {
```
Comment (timja) [inline vars/sectionSyncBranchesWithMaster.groovy:19]: this should be outside the stage, it shouldn't appear if it's not in use

File: vars/withPipeline.groovy [hunk 2]
```diff
@@ -130,6 +130,11 @@ def call(type, String product, String component, Closure body) {
             component: component,
           )
 
+          sectionSyncBranchesWithMaster(
+            branchestoSync: pipelineConfig.branchesToSyncWithMaster,
+            product: product
+          )
```
Comment (timja) [inline vars/withPipeline.groovy:136]: this is just for testing I assume?

Comment (Tyler-35) [reply vars/withPipeline.groovy:136]: Yep just testing if (!branchesToSync.isEmpty()) to see if it disables the sync if it is set to empty or does it default to ['ithc', 'demo', 'perftest']

File: vars/sectionSyncBranchesWithMaster.groovy [hunk 3]
```diff
@@ -23,19 +23,24 @@ def call(params) {
                 '''
 
                 for (branch in branchesToSync) {
-                    try {
-                        echo "Syncing branch - ${branch}"
+                    def status = sh(returnStatus: true, script: "git ls-remote --exit-code --heads origin $branch")
+                    if (status == 0) {
```
Comment (timja) [inline vars/sectionSyncBranchesWithMaster.groovy:27]:
```suggestion
                    def exists = 0
                    if (status == exists) {    
```

File: README.md [hunk 4]
```diff
@@ -674,11 +674,11 @@ The Pact broker url and other parameters are passed to these hooks as following:
 
 #### Usage
 
-The environment specific branches such as demo, ithc and perftest can be automatically synced with master branch.
+The environment specific branches such as demo, ithc and perftest are set by default to automatically be synced with master branch. If the branch doesn't exist it will not be synced.
 
-This can be achieved by using the `syncBranchesWithMaster()` method in Application, Infrastructure and Camunda pipelines. This method will be invoked in the master build and execute as the last stage in the build.
+By using the `syncBranchesWithMaster()` method in Application, Infrastructure and Camunda pipelines, you can manually override what environment branches are synced. Setting the value to 'false' will disable the sync for all branches. This method will be invoked in the master build and execute as the last stage in the build.
```
Comment (timja) [inline README.md:679]:
```suggestion
By using the `syncBranchesWithMaster()` method in Application, Infrastructure and Camunda pipelines, you can manually override what environment branches are synced. Setting the value to '[]' will disable the sync for all branches. This method will be invoked in the master build and execute as the last stage in the build.
```

**Review by timja — CHANGES_REQUESTED** [review]
i've suggested some simpler cleaner code:

File: vars/sectionSyncBranchesWithMaster.groovy [hunk 5]
```diff
@@ -10,11 +10,11 @@
 // }
 
 def call(params) {
-    def branchesToSync = params.branchestoSync
+    def branchesToSync = params.branchestoSync.exist ?: ['ithc', 'demo', 'perftest']
```
[the remark this answers is not in the document]
Comment (Tyler-35) [reply vars/sectionSyncBranchesWithMaster.groovy:13]: Try testing, all works fine except when I remove the branchesToSync variable, It doesn't default to the values we want it just doesn't run the stage.

**Review by timja — APPROVED** [review]
LGTM

**Pull request merged**